### PR TITLE
Log a warning when running SQLite in production

### LINKED ISSUE #34715: Log a warning if you run SQLite in production
### Steps to reproduce
Create a Rails app which connects to SQLite, run as production.

### Expected behavior
Log a warning that running SQLite in production is not recommended.

### Actual behavior
No warning is logged.

### System configuration
**Rails version**: Rails 5.2.2

**Ruby version**: ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin18]

### PR DESCRIPTION
**Pull request by intrip**
### Summary

Log a warning message when running SQlite in production.
The warning can be disabled by setting `config.active_record.sqlite3_production_warning=false`.

Closes #34715

### CODE REVIEW

File: activerecord/lib/active_record/railtie.rb [hunk 1]
```diff
@@ -216,10 +217,21 @@ class Railtie < Rails::Railtie # :nodoc:
...
         end
         self.configurations = Rails.application.config.database_configuration
+
         establish_connection
       end
     end
 
+    SQLITE3_PRODUCTION_WARN = "You are running SQLite in production, this is generally not recommended."\
+      " You can disable this warning by setting \"config.active_record.sqlite3_production_warning=false\"."
+    initializer "sqlite3_production_warning" do
```
Comment (SkipKayhil) [inline activerecord/lib/active_record/railtie.rb:227]:
```suggestion
    initializer "active_record.sqlite3_production_warning" do
```

I think all of the other initializers in here are prefixed

File: railties/test/application/configuration_test.rb [hunk 2]
```diff
@@ -3105,6 +3125,65 @@ class MyLogger < ::Logger
...
+        ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
+      RUBY
+      add_to_config "config.logger = MyLogRecorder.new"
+
+      app "production"
+
+      assert_match(/You are running SQLite in production, this is generally not recommended/, Rails.logger.recording)
+    end
+
+    test "doesn't log a warning when running SQlite3 in production and sqlite3_production_warning=false" do
```
Comment (eileencodes) [inline railties/test/application/configuration_test.rb:3148]:
```suggestion
    test "doesn't log a warning when running SQLite3 in production and sqlite3_production_warning=false" do
```

**Review by eileencodes — CHANGES_REQUESTED** [review]
Can you add a changelog entry?

**Comment by intrip** [comment]
@eileencodes thanks for the review! I've applied the suggestion and added the CL entry

**Review by eileencodes — APPROVED** [review]

**Comment by strzibny** [comment]
I find this change a bit strange. SQLite is a good production-like database. Shouldn't this change be about reporting mismatch between dev and production environment rather than suggesting that SQLite is not a good database?

**Comment by intrip** [comment]
You can find all the conversations + the reasons of this change in https://github.com/rails/rails/issues/34715

**Comment by strzibny** [comment]
I reread it and I am very disappointed. Seriously folks! The argument is that SQLite doesn't scale beyond a certain point? As long as it's a supported database in Rails, this change should be reverted.

**Comment by zzak** [comment]
@strzibny That's fair, and there is nothing stopping you from using SQLite in production. This warning is intended to notify people who might not be as comfortable managing that database in a production environment, if you're comfortable choosing that database as your preferred solution more power to you.

**Comment by strzibny** [comment]
@zzak The problem is that you got it backward I think.

Managing SQLite is **super simple** whereas making a production-grade MySQL/PostgreSQL deployment is _much more daunting_. A lot of beginners have trouble making even a development environment work -- I know, because in my early days I used some Mac OS app to do it for me, and it was something I didn't understand. 

Nobody shouldn't be forced to suddenly stop - and go learn about production-grade PostgreSQL deployment only to finish the deployment of their simple projects on a virtual server. They could just carry on with SQLite. It's a great first step. I and you understand that the warning is in fact wrong, but those starting would take it as a serious recommendation.

I always liked that Rails supported SQLite so that beginners (but also professionals) can make a simple app/showcase and also put it into production with very little effort. I compare it with Phoenix where we still don't have a good SQLite story and it's in fact a bit annoying. Deploying databases is not an easy feat! It requires some studying to do it right. But for SQLite? No issue. Can you manage a file? Then you can manage SQLite.

**Comment by strzibny** [comment]
I am trying to recheck the guides https://guides.rubyonrails.org/getting_started.html#creating-a-new-rails-project-installing-rails and there is nothing about SQLite not being supported for production.

I think this is about expectations. Is SQLite supported as a production database in Rails? If yes, there **cannot** be a warning about using it in production. If not, I would expect we change the docs to mention it.

Users (especially beginners) shouldn't find out the database they chose is not supported once they actually go and deploy. That's very unergnomic.

**Comment by eileencodes** [comment]
@strzibny I appreciate your perspective and concern, but the core team discussed this on the issue and came to a consensus to add this logged warning. It doesn't prevent you from using sqlite. We don't have plans to revert this at the moment.

**Comment by strzibny** [comment]
Okay. It's unfortunate, but it's up to you.

**Comment by strzibny** [comment]
@eileencodes can you please confirm or deny whether SQLite is still supported for production use in Rails? The guides mention SQLite all the time, but it's actually not clear to me if this is supported now or not.

**Comment by eileencodes** [comment]
@strzibny it's a single warning issued on boot in the production env. Nothing is _preventing_ you from using SQLIite in any environment. It is still supported the same as it was before, just with a little warning in your logs.

All the reasoning for adding this was explained in the issue. You're free to dislike it and you're free to keep using sqlite. However, please stop arguing with us about this. We are not reverting this change and I will lock this PR if you keep arguing with us.

**Pull request merged**